### ISSUE TITLE
create benchmarking code

### ISSUE THREAD

**Issue by lukasheinrich** [diana-hep/pyhf#77]
we'll be interested in scaling behaviour w.r.t number of bins / channels / systematics / etc.. as a start it should not be too hard to measure e.g. pdf evaluation time for a simple hepdata like model

e.g. this code snippet runs sets up a 2-bin likelihood, and we can just generate N-bin likelihoods with a easy loop that spits outs these JSON specs

@matthewfeickert maybe this is something for you?

```
source = {
  "binning": [2,-0.5,1.5],
  "bindata": {
    "data":    [120.0, 180.0],
    "bkg":     [100.0, 150.0],
    "bkgerr":     [10.0, 10.0],
    "sig":     [30.0, 95.0]
  }
}

from pyhf.simplemodels import hepdata_like
pdf  = hepdata_like(source['bindata']['sig'], source['bindata']['bkg'], source['bindata']['bkgerr'])
data = source['bindata']['data'] + pdf.config.auxdata

#now the call we want to benchmark:
pdf.logpdf(pdf.config.suggested_init(), data)
```

the timeit module will be useful for this

https://docs.python.org/2/library/timeit.html

**Comment by lukasheinrich** [comment]
ideally i'd have something similar to `pytest` where we can just run a series of benchmarks under certaing conditions (e.g. tensor backend etc..) which produces a summary on the cmd line or as a txt file

**Comment by matthewfeickert** [comment]
@lukasheinrich So, using [`pytest-benchmark`](http://pytest-benchmark.readthedocs.io/en/latest/index.html) it seems (to me so far at least) that we need to define a test function for every benchmark we want. I can check into this more (maybe ask the devs).

If we cant to use this (which I think we should) then to start off with some benchmarks on bins to expand upon the most important thing is just the number of bins, not necessarily the content (as long as it is somewhat reasonable), correct? Let me know if my question is confusing (it isn't phrased perhaps the best).

**Comment by matthewfeickert** [comment]
Actually I think that [this Issue](https://github.com/ionelmc/pytest-benchmark/issues/48) might help out. I'll follow up tonight.

**Comment by lukasheinrich** [comment]
@matthewfeickert yeah, i think a simple test that just clones the same single bin N times should be sufficient to observe a 0th approximation of the scaling behaviour. later on we might fuzz this out and e.g. have randomized tests

**Comment by matthewfeickert** [comment]
@lukasheinrich So I think that the way in which `pytest-benchmark` and `timeit` handle parallelization is messing things up. I wrote some dumb benchmarking code which just grabs the wall time
```python
def time_runOnePoint(n_bins):
    pdf, data = generate_pdf_and_data(n_bins)

    start_time = time.time()
    pyhf.runOnePoint(1.0, data, pdf,
                     pdf.config.suggested_init(),
                     pdf.config.suggested_bounds())
    total_wall_time = time.time() - start_time
    return total_wall_time
```
and then some wrapper functions and ran for
```python
n_bins = [1, 10, 50, 100, 200, 500, 800, 1000]
```
with the result plotted below for the different backends and optimizers.
![benchmark_times_log](https://user-images.githubusercontent.com/5142394/38436861-0eb34a04-39d6-11e8-959c-26e3e45a9396.png)

We're definitely seeing very different scaling and times then what `pytest-benchmark` was giving us.

**Comment by lukasheinrich** [comment]
hi @matthewfeickert, yeah if we cna't get pytest-benchmark to work, let's not worry about it. Nice plots. It's a bit apples to oranges in the sense that numpy is not exploiting autodiff for the minimization, but it's great to see that apparently you win significantly through it. @cranmer might be interested in seeing those.

**Comment by lukasheinrich** [comment]
@matthewfeickert so maybe you could modify this script to perhaps run a couple of thosse (so that we get a feel for variance) and print some kind of result to stdout / a file. We can then add this to the CI

**Comment by matthewfeickert** [comment]
@lukasheinrich Yeah, this was my plan. The code currently write out the test results to a JSON file, so it will be easy to instead write out the min time, max time, mean time, and variance for each number of bins.

**Comment by matthewfeickert** [comment]
It looks like I'm going to need to revisit how TensorFlow is getting called. I was noticing huge slowdowns and memory consumption and given that the relevant code is quite simple

```python
def benchmark_backend(backend, n_bins, n_runs=5):
    default_backend = pyhf.tensorlib
    pyhf.set_backend(backend)

    times = np.repeat([time_runOnePoint(bins) for bins in n_bins],
                      n_runs, axis=0)

    # Reset backend
    pyhf.set_backend(default_backend)

    return summary_stats(times)

```
I just did a simple check of how the performance changes with the number of iterations per binning choice. It seems that while NumPy and PyTorch are working as expected, TensorFlow's run time is growing as a function of `n_runs`. What is plotted below is the mean run time per binning with 1 standard deviation uncertainties and open shape points corresponding to the min and max values for the number of iterations per binning choice. The scales are different, but we see that for TensorFlow the min run time is approximately the same for `n_runs=5` and `n_runs=8` but the max time grows with `n_runs`, bringing the mean time with it.

`n_runs=5`             |  `n_runs=8`
:-------------------------:|:-------------------------:
![benchmark_times_5_log](https://user-images.githubusercontent.com/5142394/38456508-9b8c0b32-3a85-11e8-8532-45597fb861d5.png)  |  ![benchmark_times_8_log](https://user-images.githubusercontent.com/5142394/38456509-9e6b9458-3a85-11e8-93b6-8ea66dd13fb2.png)

I think this might be related to what we were seeing with `pytest-benchmark` but I'll need to investigate why TensorFlow is having problems here.

**Comment by matthewfeickert** [comment]
The above was fixed by @lukasheinrich [in Issue #102](https://github.com/diana-hep/pyhf/issues/102#issuecomment-379553003) and it has now been implemented in PR #92.

**`n_runs=5` per binning point:**
![benchmark_times_resettf_log](https://user-images.githubusercontent.com/5142394/38469179-c89763da-3b50-11e8-9037-283d8a9a319d.png)

**`n_runs=5` per binning point and testing only `[10, 100, 200]` for time considerations in CI:**
![pytest-benchmark_ci_times](https://user-images.githubusercontent.com/5142394/38469303-e16c53d2-3b52-11e8-81d5-0df618d4fcde.png)

**Comment by matthewfeickert** [comment]
@lukasheinrich I'm implementing a unit test to compare the q_μ computed by the different backends, and while SciPy and TensorFlow seem to be in good agreement it seems that PyTorch drifts (downward) from the other two.
```
# A print statement is added for this case to illustrate the values
# printing q_mu for [SciPy, TensorFlow, PyTorch]
>>> test_runOnePoint_q_mu(n_bins=50)
[22.78188789255546, 22.7816162109375, 22.73748779296875]
>>> test_runOnePoint_q_mu(n_bins=500)
[227.8188803441044, 227.814453125, 227.375]
>>> test_runOnePoint_q_mu(n_bins=800)
[364.5102065974679, 364.4912109375, 363.7998046875]
```
So I guess new questions are:
- What is happening such that PyTorch differs?
- What is a reasonable tolerance to test against?

**Comment by lukasheinrich** [comment]
@matthewfeickert that's interesting! can you make a plot of q_mu vs nbins for the backends. issues could be 

* double precision (numpy) vs single precision (but TF is also only single precision)
* different gradients between TF and scipy

the test statsistic is the ratio of two fits, it's so the problem might compound. So we should in addition also test the `unconstrained_bestfit` and `constrained_bestfit ` methods in the optimizer modules as well

**Comment by cranmer** [comment]
Could we also try the benchmarks @vincecr0ft made:

https://github.com/vincecr0ft/WorkSpaceBuilder/tree/master/WS_Builder_Numpy

**Comment by lukasheinrich** [comment]
let's close this and continue the discussion in #111
